src/Server/AspNetCore.Authorization/AuthorizeDirectiveType.cs

```diff
@@ -38,35 +38,61 @@ public class AuthorizeDirectiveType
 #if !ASPNETCLASSIC
             IAuthorizationService authorizeService = context
                 .Service<IAuthorizationService>();
+            IAuthorizationPolicyProvider policyProvider =
+                context.Service<IAuthorizationPolicyProvider>();
 #endif
             ClaimsPrincipal principal = context
                 .CustomProperty<ClaimsPrincipal>(nameof(ClaimsPrincipal));
             AuthorizeDirective directive = context.Directive
                 .ToObject<AuthorizeDirective>();
-            bool allowed = IsInRoles(principal, directive.Roles);
+            
+
+            if(!IsInRoles(principal, directive.Roles))
+            {
+                context.Result = BuildUnauthorizedError(context);
+                return;
+            }
 
 #if !ASPNETCLASSIC
-            if (allowed && !string.IsNullOrEmpty(directive.Policy))
+
+            AuthorizationPolicy policy =
+                string.IsNullOrWhiteSpace(directive.Policy)
+                    ? await policyProvider.GetDefaultPolicyAsync()
+                    : await policyProvider.GetPolicyAsync(directive.Policy);
+
+            if (policy == null)
             {
-                AuthorizationResult result = await authorizeService
-                    .AuthorizeAsync(principal, directive.Policy);
+                context.Result = BuildUnauthorizedError(context, true);
+                return;
+            }
+
+            
+            AuthorizationResult result =
+                await authorizeService.AuthorizeAsync(principal, policy);
 
-                allowed = result.Succeeded;
+            if(!result.Succeeded)
+            {
+                context.Result = BuildUnauthorizedError(context);
+                return;
             }
+            
 #endif
```

Comment:
Would be nice to have one empty line as separation between `#endif` and `await ...`.